lib/elixir/test/elixir/kernel_test.exs

```diff
@@ -359,37 +376,40 @@ defmodule KernelTest do
 
     test "with a non-literal non-escaped compile-time range in guards" do
       message = "non-literal range in guard should be escaped with Macro.escape/2"
-      assert_raise ArgumentError, message, fn ->
-        Code.eval_string """
-        defmodule InErrors do
-          range = 1..3
-          def foo(x) when x in unquote(range), do: :ok
-        end
-        """
+      assert_eval_string_error ArgumentError, message, """
+      defmodule InErrors do
+        range = 1..3
+        def foo(x) when x in unquote(range), do: :ok
       end
+      """
     end
 
     test "with a non-compile-time range in guards" do
       message = ~r/invalid args for operator "in", .* got: :hello/
-      assert_raise ArgumentError, message, fn ->
-        Code.eval_string """
-        defmodule InErrors do
-          def foo(x) when x in :hello, do: :ok
-        end
-        """
+      assert_eval_string_error ArgumentError, message, """
+      defmodule InErrors do
+        def foo(x) when x in :hello, do: :ok
       end
+      """
     end
 
     test "with a non-compile-time list cons in guards" do
-      message = ~r/invalid args for operator "in", .* got: list\(\)/
-      assert_raise ArgumentError, message, fn ->
-        Code.eval_string """
-        defmodule InErrors do
-          def list, do: [1]
-          def foo(x) when x in [1 | list()], do: :ok
-        end
-        """
+      message = ~r"cannot invoke local list/0 inside guard, called as: list()"
+      assert_eval_string_error CompileError, message, """
+      defmodule InErrors do
+        def list, do: [1]
+        def foo(x) when x in [1 | list()], do: :ok
+      end
+      """
+    end
+
+    test "with a compile-time non-list in tail in guards" do
+      message = ~r"cannot invoke remote function :lists.member/2 inside guard"
+      assert_eval_string_error CompileError, message, """
```

Comment:
@josevalim I'm not a fan of this error message, but I'm not sure we can do something better.